File: app/src/main/java/de/nicidienase/chaosflix/entities/streaming/Room.java
package de.nicidienase.chaosflix.entities.streaming;

import android.os.Parcel;
import android.os.Parcelable;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by felix on 23.03.17.
 */

public class Room implements Parcelable{
	String slug;
	String shedulename;
	String thumb;
	String link;
	String display;
	List<Stream> streams;

	protected Room(Parcel in) {
		slug = in.readString();
		shedulename = in.readString();
		thumb = in.readString();
		link = in.readString();
		display = in.readString();
	}

	public static final Creator<Room> CREATOR = new Creator<Room>() {
		@Override
		public Room createFromParcel(Parcel in) {
			return new Room(in);
		}

		@Override
		public Room[] newArray(int size) {
			return new Room[size];
		}
	};

	@Override
	public int describeContents() {
		return 0;
	}

	@Override
	public void writeToParcel(Parcel dest, int flags) {
		dest.writeString(slug);
		dest.writeString(shedulename);
		dest.writeString(thumb);
		dest.writeString(link);
		dest.writeString(display);
	}

	public String getSlug() {
		return slug;
	}

	public void setSlug(String slug) {
		this.slug = slug;
	}

	public String getShedulename() {
		return shedulename;
	}

	public void setShedulename(String shedulename) {
		this.shedulename = shedulename;
	}

	public String getThumb() {
		return thumb;
	}

	public void setThumb(String thumb) {
		this.thumb = thumb;
	}

	public String getLink() {
		return link;
	}

	public void setLink(String link) {
		this.link = link;
	}

	public String getDisplay() {
		return display;
	}

	public void setDisplay(String display) {
		this.display = display;
	}

	public List<Stream> getStreams() {
		return streams;
	}

	public void setStreams(List<Stream> streams) {
		this.streams = streams;
	}
}

File: common/src/main/java/de/nicidienase/chaosflix/common/mediadata/entities/streaming/Room.kt
package de.nicidienase.chaosflix.common.mediadata.entities.streaming

import android.os.Parcel
import android.os.Parcelable
import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import java.util.*
import kotlin.collections.HashMap

@JsonIgnoreProperties(ignoreUnknown = true)
data class Room(var slug: String,
                var schedulename: String,
                var thumb: String,
                var link: String,
                var display: String,
                var talks: Map<String,StreamEvent>?,
                var streams: List<Stream>) : Parcelable {


	protected constructor(input: Parcel) : this(
			slug = input.readString() ?: "",
			schedulename = input.readString() ?: "",
			thumb = input.readString() ?: "",
			link = input.readString() ?: "",
			display = input.readString() ?: "",
			talks = readMap(input),
			streams = input.createTypedArrayList<Stream>(Stream.CREATOR).filterNotNull())


	override fun describeContents(): Int {
		return 0
	}

	override fun writeToParcel(dest: Parcel, flags: Int) {
		dest.writeString(slug)
		dest.writeString(schedulename)
		dest.writeString(thumb)
		dest.writeString(link)
		dest.writeString(display)
		val talkKeys = talks?.keys?.toTypedArray()
		dest.writeStringArray(talkKeys)
		val talks = talkKeys?.map { talks?.get(it) }?.toTypedArray()
		dest.writeTypedArray(talks,0)
		dest.writeTypedList(streams)
	}


	companion object CREATOR : Parcelable.Creator<Room> {


		override fun createFromParcel(`in`: Parcel): Room {
			return Room(`in`)
		}

		override fun newArray(size: Int): Array<Room?> {
			return arrayOfNulls(size)
		}

		fun readMap(input: Parcel): MutableMap<String, StreamEvent>? {
			val keys = input.createStringArray()
			val urls = input.createTypedArray(StreamEvent.CREATOR)
			if(keys == null || urls == null){
				return null
			}
			val result = HashMap<String, StreamEvent>()
			for(i in 0 until keys.size - 1){
				val key = keys[i]
				val value = urls[i]
				if(key != null && value != null){
					result.put(key, value)
				}
			}
			return result
		}

		val dummyObject: Room
			get() {
				val dummy = Room(
						"dummy_room",
						"Dummy Room",
						"https://static.media.ccc.de/media/unknown.png",
						"",
						"Dummy Room",
						HashMap(),
						ArrayList())
				dummy.streams = listOf(Stream.dummyObject)
				return dummy
			}

	}
}